pkg/migration-tool/planner/block.go

```diff
@@ -21,24 +21,89 @@ type Block struct {
+	mint, maxt int64
+}
+
+// initStores initializes the stores.
+func (b *Block) initStores() {
+	var (
+		proceed   = b.mint
+		increment = (b.maxt - b.mint) / int64(b.numStores)
+	)
+	for storeIndex := 0; storeIndex < b.numStores; storeIndex++ {
+		b.stores[storeIndex] = store{
+			id:   storeIndex,
+			mint: proceed,
+			maxt: proceed + increment,
+		}
+		proceed += increment
+	}
+	b.stores[b.numStores-1].maxt = b.maxt // To ensure that we cover the entire range, which may get missed due to integer division in increment.
+}
+
+// Fetch starts fetching the samples from remote read storage based on the matchers. It takes care of concurrent pulls as well.
+func (b *Block) Fetch(ctx context.Context, client *utils.Client, mint, maxt int64, matchers []*labels.Matcher) ([]*prompb.TimeSeries, error) {
+	var (
+		totalRequests         = b.numStores
+		readRequests          = make([]*prompb.Query, totalRequests)
+		readResponses         = make([]*prompb.QueryResult, totalRequests)
+		contexts              = make([]context.Context, totalRequests)
+		cancelFuncs           = make([]context.CancelFunc, totalRequests)
+		timeRangeMinutesDelta = (maxt - mint) / time.Minute.Milliseconds()
+	)
+	for i := 0; i < totalRequests; i++ {
+		readRequest, err := utils.CreatePrombQuery(b.stores[i].mint, b.stores[i].maxt, matchers)
+		if err != nil {
+			return nil, fmt.Errorf("create promb query: %w", err)
+		}
+		readRequests[i] = readRequest
+		cctx, cancelFunc := context.WithCancel(ctx)
+		contexts[i] = cctx
+		cancelFuncs[i] = cancelFunc
+	}
+	var (
+		doneChan = make(chan *utils.PrompbResponse, totalRequests)
+		errChan  = make(chan error, totalRequests)
+	)
+	for i := 0; i < totalRequests; i++ {
+		// Initiate concurrent fetching of data.
+		go client.Read(contexts[i], readRequests[i], i, fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta), doneChan, errChan)
+	}
+	var (
+		bytesCompressed   int
+		bytesUncompressed int
+	)
+	for i := 0; i < totalRequests; i++ {
+		select {
+		case resp := <-doneChan:
+			bytesCompressed += resp.NumBytesCompressed
+			bytesUncompressed += resp.NumBytesUncompressed
+			readResponses[resp.ID] = resp.Result
+		case err := <-errChan:
+			for _, cancelFnc := range cancelFuncs {
+				cancelFnc()
+			}
+			return nil, fmt.Errorf("executing client-read: %w", err)
+		}
 	}
-	result, bytesCompressed, bytesUncompressed, err := client.Read(context, readRequest, fmt.Sprintf("%s | time-range: %d mins", b.pbarDescriptionPrefix, timeRangeMinutesDelta))
-	if err != nil {
-		return nil, fmt.Errorf("executing client-read: %w", err)
+	close(doneChan)
+	close(errChan)
+	// Combine the stores.
+	var result []*prompb.TimeSeries
+	for storeIndex := 0; storeIndex < b.numStores; storeIndex++ {
+		result = append(result, readResponses[storeIndex].Timeseries...)
 	}
-	b.timeseries = result.Timeseries
+	b.timeseries = result
```

Comment:
Was there a reason why the result is cached in the block? (since I see it also returned from this func)

Reply:
This is to return the combined stores (as blocks) to the writer. But yes, we should make the stores `nil` after this operation to save memory.

Reply:
but what about b.timeseries?
result is assigned to it here, and result is returned from the method, why is b.timeseries needed?

Reply:
`b.timeseries` is used to return the timeseries in the writer (https://github.com/timescale/promscale/blob/5f87ccd9f09a8aaa2549e441d79120e4f28f1e98/pkg/migration-tool/planner/block.go#L62). The result is returned from here only to make sure that this result is not empty.

Reply:
But yes, this does points out an important point. The function does not need to return such hefty values that will slow the registers. Since this returns only determins whether its empty or not, we can return `true` or `false`, for `containsTs` and this would be far efficient. Thank you for pointing this out.